source/common/decompressor/zlib_decompressor_impl.cc

```diff
@@ -0,0 +1,68 @@
+#include "common/decompressor/zlib_decompressor_impl.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Decompressor {
+
+ZlibDecompressorImpl::ZlibDecompressorImpl() : ZlibDecompressorImpl(4096){};
+
+ZlibDecompressorImpl::ZlibDecompressorImpl(uint64_t chunk_size)
+    : chunk_{chunk_size}, initialized_{false}, output_char_ptr_(new unsigned char[chunk_size]),
+      zstream_ptr_(new z_stream(), [](z_stream* z) {
+        inflateEnd(z);
+        delete z;
+      }) {
+  zstream_ptr_->zalloc = Z_NULL;
+  zstream_ptr_->zfree = Z_NULL;
+  zstream_ptr_->opaque = Z_NULL;
+  zstream_ptr_->avail_out = chunk_;
+  zstream_ptr_->next_out = output_char_ptr_.get();
+}
+
+void ZlibDecompressorImpl::init(int8_t window_bits) {
+  ASSERT(initialized_ == false);
+  const int result = inflateInit2(zstream_ptr_.get(), static_cast<int>(window_bits));
+  RELEASE_ASSERT(result >= 0);
+  initialized_ = true;
+}
+
+uint64_t ZlibDecompressorImpl::checksum() { return zstream_ptr_->adler; }
+
+void ZlibDecompressorImpl::decompress(const Buffer::Instance& input_buffer,
+                                      Buffer::Instance& output_buffer) {
+  const uint64_t num_slices = input_buffer.getRawSlices(nullptr, 0);
+  Buffer::RawSlice slices[num_slices];
+  input_buffer.getRawSlices(slices, num_slices);
+
+  for (const Buffer::RawSlice& input_slice : slices) {
+    zstream_ptr_->avail_in = input_slice.len_;
+    zstream_ptr_->next_in = static_cast<Bytef*>(input_slice.mem_);
+    while (inflateNext()) {
+      if (zstream_ptr_->avail_out == 0) {
+        output_buffer.add(static_cast<void*>(output_char_ptr_.get()),
+                          chunk_ - zstream_ptr_->avail_out);
+        output_char_ptr_.reset(new unsigned char[chunk_]);
+        zstream_ptr_->avail_out = chunk_;
+        zstream_ptr_->next_out = output_char_ptr_.get();
+      }
+    }
+  }
+
+  output_buffer.add(static_cast<void*>(output_char_ptr_.get()), chunk_ - zstream_ptr_->avail_out);
```

Comment:
Is it correct to say that `chunk_ - zstream_ptr_->avail_out` might be 0? Does it need to be special cased?

Reply:
It doesn't make sense to call it when has no output.